### PR TITLE
Provide support to Document API for DynamoDB

### PR DESCRIPTION
**Pull request by dearrudam**
Ref: https://github.com/eclipse/jnosql/issues/380

Steps completed:

- [x] DocumentConfiguration implementation;
- [x] DocumentManagerFactory implementation;
- [x] DocumentEntity converter to DynamoDB model;
- [x] DocumentManager implementation:
   - [x] insert
   - [x] update
   - [x] delete
   - [x] select
   - [x] count by table name
   - [x] count by query
- [x] Mapping implementations
   - [x] CDI Support
   - [x] JNoSQLDocumentTemplate implementation
- [x] Create a readme explaining the configurations
- [x] Create document API for configuration
- [x] Update changelog

### CODE REVIEW

**Comment by genie-jnosql** [comment]
Can one of the admins verify this patch?

File: jnosql-dynamodb/pom.xml [hunk 1]
```diff
@@ -57,6 +57,12 @@
             <artifactId>dynamodb-enhanced</artifactId>
             <version>${dynamodb.version}</version>
         </dependency>
+        <dependency>
```
Comment (otaviojava) [inline jnosql-dynamodb/pom.xml:60]: Hey Max, how about including this as JNoSQL parent?
This might be valuable to everyone.

Comment (dearrudam) [reply jnosql-dynamodb/pom.xml:60]: @otaviojava, sorry, I didn't understand... I'm still working on this PR... do you think that's a good idea to merge this PR as it is now into the main branch?

Comment (dearrudam) [reply jnosql-dynamodb/pom.xml:60]: Oh... I think that I understand now, @otaviojava !!! You mean to add the DataFaker at JNoSQL parent POM!!! When I saw it on my smartphone, I didn't realize you were talking about it... LOL

Comment (dearrudam) [reply jnosql-dynamodb/pom.xml:60]: I'm gonna add this one to the JNoSQL parent POM!

Comment (dearrudam) [reply jnosql-dynamodb/pom.xml:60]: It's done!: https://github.com/eclipse/jnosql/pull/473

**Comment by dearrudam** [comment]
@otaviojava, please, review this PR?!

**Comment by otaviojava** [comment]
@dearrudam I have plan to release a new version this week with Oracle NoSQL.
If you need more time we could move this one to the next release.

Please, let me know if you need any help on this.

File: jnosql-dynamodb/pom.xml [hunk 2]
```diff
@@ -44,6 +52,12 @@
             <artifactId>dynamodb</artifactId>
             <version>${dynamodb.version}</version>
         </dependency>
+        <dependency>
+            <groupId>net.datafaker</groupId>
```
[the remark this answers is not in the document]
Comment (dearrudam) [reply jnosql-dynamodb/pom.xml]: Yeah... I've rebased the branch... it looks nice now!

**Comment by dearrudam** [comment]
@otaviojava well, I don't see problems adding this feature into this release this week, unless you think that it's better... btw, I think more things will be improved along the way...

**Comment by otaviojava** [comment]
@dearrudam could you rebase and test this branch?
Could we move forward and merge this branch?

File: jnosql-dynamodb/pom.xml [hunk 3]
```diff
@@ -34,6 +34,14 @@
             <groupId>org.eclipse.jnosql.mapping</groupId>
             <artifactId>jnosql-mapping-key-value</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.jnosql.communication</groupId>
+            <artifactId>jnosql-communication-document</artifactId>
```
Comment (otaviojava) [inline jnosql-dynamodb/pom.xml:39]: We don't have a document project anymore; could you update it?

Comment (otaviojava) [inline jnosql-dynamodb/pom.xml:39]: Please, remove this dependency. The document already has it.

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        return dynamoDbClient;
+    }
+
+    @Override
+    public String name() {
+        return database;
+    }
+
+    @Override
+    public DocumentEntity insert(DocumentEntity documentEntity) {
```
Comment (otaviojava) [inline jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java:99]: we merged and the DocumentEntity became CommunicationEntity

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    @Override
+    public Iterable<DocumentEntity> update(Iterable<DocumentEntity> entities) {
+        requireNonNull(entities, "entities is required");
+        return StreamSupport.stream(entities.spliterator(), false)
+                .map(this::update)
+                .toList();
+    }
+
+    @Override
+    public void delete(DocumentDeleteQuery documentDeleteQuery) {
```
Comment (otaviojava) [inline jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java:259]: You can use the MongoDB driver as a sample. It was replaced to `DeleteQuery`

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    @Override
+    public void close() {
+        this.dynamoDbClient.close();
+    }
+
+    @Override
+    public Stream<DocumentEntity> partiQL(String query) {
+        return partiQL(query,new Object[0]);
```
Comment (otaviojava) [inline jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java:323]: Once it is varags, why not leave it empty?

```suggestion
        return partiQL(query);
```

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * Maximillian Arruda
+ */
+
+package org.eclipse.jnosql.databases.dynamodb.communication;
+
+import org.eclipse.jnosql.communication.Settings;
+import org.eclipse.jnosql.communication.document.DocumentConfiguration;
+
+public class DynamoDBDocumentConfiguration extends DynamoDBConfiguration
+        implements DocumentConfiguration {
```
Comment (otaviojava) [inline jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentConfiguration.java:22]: This one becames `DatabaseConfiguration`

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentManager.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import org.eclipse.jnosql.communication.document.DocumentEntity;
+import org.eclipse.jnosql.communication.document.DocumentManager;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+
+import java.util.stream.Stream;
+
+/**
+ * A document manager interface for DynamoDB database operations.
+ */
+public interface DynamoDBDocumentManager extends DocumentManager {
```
Comment (otaviojava) [inline jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentManager.java:27]: DatabaseManager

File: jnosql-dynamodb/pom.xml (hunk 3 above)
Comment (dearrudam) [reply jnosql-dynamodb/pom.xml]: Fixed!

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java (hunk 4 above)
Comment (dearrudam) [reply jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java]: Fixed!

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+                    .build());
+
+            var tableRequest = DescribeTableRequest.builder().tableName(tableName).build();
+            var waiterResponse = waiter.waitUntilTableExists(tableRequest);
+            return waiterResponse.matched().response().orElseThrow();
+        }
+    }
+
+    private StreamSpecification defaultStreamSpecificationFor() {
+        return null;
```
[the remark this answers is not in the document]
Comment (dearrudam) [reply jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java]: Yeah! It was removed!

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java (hunk 5 above)
Comment (dearrudam) [reply jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java]: Fixed!

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+        select(selectQueryBuilder.build()).forEach(
+                documentEntity ->
+                        dynamoDbClient().deleteItem(DeleteItemRequest.builder()
+                                .tableName(documentDeleteQuery.name())
+                                .key(getItemKey(documentEntity))
+                                .build()));
+    }
+
+    @Override
+    public Stream<DocumentEntity> select(DocumentQuery documentQuery) {
```
[the remark this answers is not in the document]
Comment (dearrudam) [reply jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java]: Fixed!

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java (hunk 6 above)
Comment (dearrudam) [reply jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DefaultDynamoDBDocumentManager.java]: Yeah! I fixed it!

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentConfiguration.java (hunk 7 above)
Comment (dearrudam) [reply jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentConfiguration.java]: Fixed!

File: jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentManager.java (hunk 8 above)
Comment (dearrudam) [reply jnosql-dynamodb/src/main/java/org/eclipse/jnosql/databases/dynamodb/communication/DynamoDBDocumentManager.java]: Fixed!

**Comment by dearrudam** [comment]
@otaviojava I fixed the classes, so, feel free to review it!

Also, I think that it's needed to take a look at the usage of the following annotations, because the AWS DynamoDB users are using the SINGLE TABLE strategy by default and I think this behavior can be achieved by using these annotations:

- `@Inheritance` 
- `@DiscriminatorColumn` 
- `@DiscriminatorValue`

**Pull request merged**